attribution/attribution_metrics.py

```diff
@@ -0,0 +1,171 @@
+import math
+from typing import List, Tuple
+
+import numpy as np
+import openai
+from sklearn.metrics.pairwise import cosine_similarity
+from transformers import PreTrainedModel, PreTrainedTokenizer
+
+
+def token_prob_difference(
+    initial_logprobs: openai.types.chat.chat_completion.ChoiceLogprobs,
+    perturbed_logprobs: openai.types.chat.chat_completion.ChoiceLogprobs,
+) -> Tuple[float, List[str], np.ndarray]:
+    # Extract token and logprob from initial_logprobs
+    initial_token_logprobs = [
+        (logprob.token, logprob.logprob) for logprob in initial_logprobs.content
+    ]
+    initial_tokens = [content.token for content in initial_logprobs.content]
+
+    # Create a list of dictionaries with token and top logprobs from perturbed_logprobs
+    perturbed_token_logprobs_list = [
+        {
+            top_logprob.token: top_logprob.logprob
+            for top_logprob in token_content.top_logprobs
+        }
+        for token_content in perturbed_logprobs.content
+    ]
+
+    # Probability change for each input token
+    prob_difference_per_token = np.zeros(len(initial_tokens))
+    NEAR_ZERO_PROB = -100  # Logprob constant for near zero probability
+
+    # Calculate the absolute difference in probabilities for each token
+    for i, initial_token in enumerate(initial_token_logprobs):
+        perturbed_token_logprobs = (
+            perturbed_token_logprobs_list[i]
+            if i < len(perturbed_token_logprobs_list)
+            else {}
+        )
+        perturbed_logprob = perturbed_token_logprobs.get(
+            initial_token[0], NEAR_ZERO_PROB
+        )
+        prob_difference_per_token[i] = abs(
+            math.exp(initial_token[1]) - math.exp(perturbed_logprob)
+        )
+
+    # Note: Different length outputs shift the mean upwards. This may or may not be desired behaviour.
+    return prob_difference_per_token.mean(), initial_tokens, prob_difference_per_token
+
+
+def token_displacement(
+    initial_logprobs: openai.types.chat.chat_completion.ChoiceLogprobs,
+    perturbed_logprobs: openai.types.chat.chat_completion.ChoiceLogprobs,
+) -> Tuple[float, List[str], np.ndarray]:
+    initial_tokens = [content.token for content in initial_logprobs.content]
+    perturbed_top_tokens = [
+        [top_logprob.token for top_logprob in token_content.top_logprobs]
+        for token_content in perturbed_logprobs.content
+    ]
+
+    # Token displacement for each initially predicted token
+    displacement_per_token = np.zeros(len(initial_tokens))
+    MAX_TOKEN_DISPLACEMENT = 20
+    for i, token in enumerate(initial_tokens):
+        if i < len(perturbed_top_tokens) and token in perturbed_top_tokens[i]:
+            displacement_per_token[i] = abs(i - perturbed_top_tokens[i].index(token))
```

Comment:
Not sure about `i - perturbed_top_tokens[i]` - is that right?

Reply:
You're right, thanks! 

The token displacement should just be `perturbed_top_tokens[i].index(token)`